internal/application/services.go

```diff
@@ -15,6 +17,173 @@ import (
+					}
+					fetched[appName] = app
+				}
+				result[serviceName] = append(result[serviceName], *app)
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// BoundAppsNamesFor is a specialization of BoundAppsNames after it, to retrieve the names
+// of the applications bound to a single service, specified by name.
+
+func BoundAppsNamesFor(ctx context.Context, cluster *kubernetes.Cluster, namespace, serviceName string) ([]string, error) {
+	result := []string{}
+
+	// locate service bindings managed by epinio applications
+	selector := EpinioApplicationAreaLabel + "=service"
+	selector += ",app.kubernetes.io/component=application"
+	selector += ",app.kubernetes.io/managed-by=epinio"
+
+	appBindings, err := cluster.Kubectl.CoreV1().Secrets(namespace).List(ctx,
+		metav1.ListOptions{
+			LabelSelector: selector,
+		})
+	if err != nil {
+		return result, err
+	}
+
+	// Instead of building a full inverted map from service names to app names here we
+	// filter on the service name to generate just that slice of app names.
+
+	for _, binding := range appBindings.Items {
+		for boundServiceName := range binding.Data {
+			if boundServiceName == serviceName {
+				appName := binding.ObjectMeta.Labels["app.kubernetes.io/name"]
+				result = append(result, appName)
+				break
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// BoundAppsNames returns a map from the service names of services in the specified namespace,
+// to the names of the applications they are bound to. For an empty namespace the map
+// contains the services in all namespaces, and the keys are a combination of namespace
+// name and service name, to distinguish same-named services in different namespaces (See
+// `ServiceKey` below). The application names never contain namespace information, as they
+// are always in the same namespace as the service referencing them.
+
+func BoundAppsNames(ctx context.Context, cluster *kubernetes.Cluster, namespace string) (map[string][]string, error) {
+
+	result := map[string][]string{}
+
+	// locate service bindings managed by epinio applications.
+	selector := EpinioApplicationAreaLabel + "=service"
+	selector += ",app.kubernetes.io/component=application"
+	selector += ",app.kubernetes.io/managed-by=epinio"
+
+	appBindings, err := cluster.Kubectl.CoreV1().Secrets(namespace).List(ctx,
+		metav1.ListOptions{
+			LabelSelector: selector,
+		})
+	if err != nil {
+		return result, err
+	}
+
+	if namespace == "" {
+		// services are collected across all namespaces.
+		// Key the map by service and namespace!
+		// Because services of the same name can exist in
+		// different namespaces, and different binding states.
+
+		for _, binding := range appBindings.Items {
+			appName := binding.ObjectMeta.Labels["app.kubernetes.io/name"]
+			namespace := binding.ObjectMeta.Labels["app.kubernetes.io/part-of"]
+
+			for serviceName := range binding.Data {
+				key := ServiceKey(serviceName, namespace)
+				result[key] = append(result[key], appName)
+			}
+		}
+	} else {
```

Comment:
Is there a reason why we can't always use a key that has the namespace in? I mean, even if the caller asked only for services of one namespace, it obviously has that namespace at hand and can construct a key for each service to look it up. It would make this code but also the caller simpler if the result of this method was consistent.

Reply:
True. Time spent in encoding/decoding keys should be small. Will do.

Reply:
commit 858aeddc147ade02f2e1c91012b1809ac3f6989a